core-tests/shared/src/test/scala/zio/ZIOParCombinatorsSpec.scala

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package zio
+
+import zio.test._
+import zio.test.Assertion._
+
+object ZIOParCombinatorsSpec
+    extends ZIOBaseSpec(
+      suite("ZIO *Par combinators")(
+        testM("foreachPar_ happy-path") {
+          val as = Seq(1, 2, 3, 4, 5)
+          for {
+            ref <- Ref.make(Seq.empty[Int])
+            _   <- ZIO.foreachPar_(as)(a => ref.update(_ :+ a))
+            rs  <- ref.get
+          } yield assert(rs.sorted, equalTo(as))
+        },
+        testM("foreachPar_ huge") {
+          val as = (1 to 1600).toList
+          for {
+            ref <- Ref.make(List.empty[Int])
+            _   <- ZIO.foreachPar_(as)(a => ref.update(_ :+ a))
+            rs  <- ref.get
+          } yield assert(rs.sorted, equalTo(as))
+        },
+        testM("foreachPar_ cancels other computations and propagates errors") {
+          val boom = new Exception("Boom!")
+          for {
+            semaphore <- Semaphore.make(0)
```

Comment:
Could I bother you to use a `Ref[Int]` instead of a semaphore? We're going to get rid of `acquireN` and `releaseN`.

Reply:
I used `Semaphore` because `onInterrupt` might be called only after we already called `counter.get` where `counter` is `Ref[Int]`. We might be able to replace it with some sort of `eventually` call on `Ref[Int]`, but I'm not really sure if it is any better.

Reply:
Ah I see. Ok. I'm fine with that. Let's tackle this when we change Semaphore.